core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +219,21 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if left side is smaller than the right by more than SIZE_DIFFERENCE_THRESHOLD times.
+        // We use 8x factor because the reduction in size from aggregation and filtering
+        // may not be fully accounted here.
+        // We use getApproximateOutputSizeInBytes instead of getSourceTablesSizeInBytes to account for the effects of
+        // the operators between join and table scan as much as possible when comparing the size of the join sides.
```

Comment:
I would put it differently. Along the lines:
1. We couldn't estimate join with `reorder join` so some estimates are missing
2. For selecting broadcast join we must be more conservative, because..., so for broadcast join we don't account for operators that could reduce data
3. For flipping join sides we don't have be that conservative (so we account for operators that potentially reduce data), because if we make a wrong decision, it's not as bad as choosing broadcast join.

Reply:
> For flipping join sides we don't have be that conservative

If we look at tpch/q09 in the results with this change it's peak memory changes from 549GB to 110GB and runtime from 102 sec to 75 sec. So there is a significant change to the query's resource requirements even with flipping join sides on a repartitioned join. Since we could potentially make a wrong decision of similar magnitude, I think that justifies being conservative with this decision.